### ISSUE TITLE
Encoding `Data.BigInteger`s encodes as string

### ISSUE THREAD

**Issue by emiflake** [input-output-hk/plutus#3585]
## Area

- [ ] *Plutus Foundation* Related to the GHC plugin, Haskell-to-Plutus compiler, on-chain code
- [x] *Plutus Application Framework* Related to the Plutus application backend (PAB), emulator, Plutus libraries
- [ ] *Marlowe* Related to Marlowe
- [ ] *Other* Any other topic (Playgrounds, etc.)

## Summary

Not entirely sure if this is per se something that can be fixed in a convenient way, but it should be documented at least.

`BigInteger`s encode wrongly as strings by default when using Data.BigInteger and this causes aeson to fail to decode. 

## Steps to reproduce

A specific example that a user is quite likely to encounter: 

The `Wallet` type (as generated by PSGenerator.Common) will create a newtype containing a `BigInteger`. Whenever we `encode` this using `Foreign.Class`, and consequently with `stringify`, we will face a `"2"`, for `Wallet { getWallet: BigInteger.fromInt 2 }`. Which causes aeson to complain on the PAB API decode. The problem lies in `JSON.stringify` for `bignumber.js`.

## Expected behavior

`JSON.stringify` should work correctly on a `Foreign`-encoded BigInteger.

## Workarounds

Using [json-bigint](https://www.npmjs.com/package/json-bigint)'s `stringify` seems to circumvent the problem.

## Versions

- `bignumber.js ^9.0.1`
- `web-common(-plutus) from rev 5cdd2c3d708bf4c33514681dee096da6463273b7`
- `node v14.15.4`
- `purs 0.13.8`

**Comment by luigy** [comment]
@emiflake thanks for the report. I did a test in `plutus-playground-client` and I'm getting a different behavior with `encodeJSON` if that's an option. Here is what I tried roughly:
```purs
import Wallet.Emulator.Wallet
import Data.BigInteger as BI
import Foreign.Generic (encodeJSON)

let
  w :: Wallet
  w = wrap ({ getWallet : BI.fromInt 1 })
in [ div_ [ text $ encodeJSON $ BI.fromInt 100]
   , div_ [ text $ encodeJSON w ]
   ]
```
![2021-07-20-173111_152x70_scrot](https://user-images.githubusercontent.com/817138/126398222-2239306b-8421-4204-86aa-d2ff6978aaaa.png)

**Comment by emiflake** [comment]
Yeah, it appears that encoding with `encodeJSON` _does_ produce correct stringified results (similar to my approach). Manually `encode`-ing and then using `JSON.stringify` doesn't (this is what for example Affjax does when you pass it a `Json`)
![image](https://user-images.githubusercontent.com/7753922/126470663-5acda4a7-cfe1-483a-bde2-4a5f37898a44.png)

**Comment by silky** [comment]
If it helps, I think this is related to a quirk in JSON; for example if you just have a large integer, then tools like `jq` choose to interpret it as a floating-point:

```
echo '{"a":9128730192384570193485710934857}' | jq
{
  "a": 9.12873019238457e+30
}
```

which is frustrating :) So in some sense it really isn't safe to encode BigIntegers as "integers" in JSON at all; they probably need to be wrapped in strings, so maybe that's why some libraries are making this choice.

**Comment by emiflake** [comment]
Yeah, I understand. RFC7159 says _"This specification allows implementations to set limits on the range and precision of numbers accepted."_ So it all comes down to what the specific implementations support. Aeson supports decoding large integers into `Integer`, and since we _can_ get PureScript to produce those too, it's fine for those purposes. It's unfortunate to stringly-type BigIntegers just out of fear for other implementations not supporting them. I guess in the case of PureScript we seem to have "reasonable" way to choose which encoding we prefer.
